File: Unit 2-Data Structure/Stack/As Array/Stack.java
class AStack<T>
{
    private static final int MAX_SIZE = 100;
    private T stack[];
    private int top;
    
    @SuppressWarnings("unchecked")
    public AStack()
    {
        stack = (T[]) new Object[MAX_SIZE];
        top = -1;
    }

    public boolean push(T value) 
    {
        if (top >= MAX_SIZE)
        {
            System.out.println("Stack OverFlow");
            return false;
        }
        else 
        {
            top++;
            stack[top] = value;
            return true;
        }
    }

    public T pop()
    {
        if (top < 0)
        {
            System.out.println("Stack UnderFlow");
            return null; // Adjust return type based on your requirements
        } 
        else
        {
            T item = stack[top];
            top--;
            return item;
        }
    }

    public boolean isEmpty()
    {
        return top < 0;
    }

    public boolean isFull()
    {
        return top >= MAX_SIZE - 1;
    }

    public T peek()
    {
        if (top < 0)
        {
            System.out.println("Stack UnderFlow");
            return null; // Adjust return type based on your requirements
        }
        else
        {
            return stack[top];
        }
    }
}


class Stack
{
    public static void main(String[] args)
    {
        AStack<Integer> myStack = new AStack<>();

        for (int i = 0; i <= 10; i++)
        {
            myStack.push(i);
        }

        System.out.println("Top element: " + myStack.peek());
        System.out.println("Is the stack empty? " + (myStack.isEmpty() ? "Yes" : "No"));
        System.out.println("Is the stack full? " + (myStack.isFull() ? "Yes" : "No"));
    }
}

File: Unit 2-Data Structure/Stack/As Array/Stack.py
class AStack:
    MAX_SIZE = 100

    def __init__(self):
        self.stack = [None] * self.MAX_SIZE
        self.top = -1

    def push(self, value):
        if self.top >= self.MAX_SIZE - 1:
            print("Stack Overflow")
            return False
        else:
            self.top += 1
            self.stack[self.top] = value
            return True

    def pop(self):
        if self.top < 0:
            print("Stack Underflow")
            return None  # Adjust return type based on your requirements
        else:
            item = self.stack[self.top]
            self.top -= 1
            return item

    def is_empty(self):
        return self.top < 0

    def is_full(self):
        return self.top >= self.MAX_SIZE - 1

    def peek(self):
        if self.top < 0:
            print("Stack Underflow")
            return None  # Adjust return type based on your requirements
        else:
            return self.stack[self.top]

if __name__ == "__main__":
    my_stack = AStack()

    for i in range(11):
        my_stack.push(i)

    print("Top element:", my_stack.peek())
    print("Is the stack empty?", "Yes" if my_stack.is_empty() else "No")
    print("Is the stack full?", "Yes" if my_stack.is_full() else "No")
